Update language script to run on Windows

gudnimg opened the pull request:
* Create new script `update-po.py` which can be run instead of `update_po.sh`
* Create new scripts to replace `.sh` files:
    *  `update-po.py` can replace `update-po.sh`
    * `update-pot.py` can replace `update-pot.sh`
* Update `lang-extract.py` to run smoothly on Windows by handling regex pattern

gudnimg committed:
languages: create new script update_po.py

The script should do the same work as when one run update_po.sh
The Python script can be run on Windows too.

gudnimg committed:
languages: update lang-extract.py

Fix an issue where it does not run on Windows.
When the input argument is regex, we must first search for the files
using said regex pattern and then creating a list of files.
When opening the files, the encoding needs to be utf-8.

gudnimg committed:
languages: Add new script update-pot.py

gudnimg commented:
I'm converting this PR to a draft so it can't be merged. I'd like to spend a little more time on this to polish it.

3d-gussner commented:
> I'm converting this PR to a draft so it can't be merged. I'd like to spend a little more time on this to polish it.

I can approve that state of today. It is working on Linux and Windows (after reinstalling Python and `pip install poedit regex` libs.

gudnimg commented:
Ran some unit tests to try to recreate the Exception, but I did not see it so I'm opening this PR again for review.

gudnimg committed:
Allow languages to be updated from any directory

Now these commands will work:
python .\lang\update-po.py
python .\lang\update-pot.py

When current working directory is
.\Prusa-Firmware\

Previously, you'd need to call the
script within the lang folder.

gudnimg committed:
Add requirements.txt

gudnimg committed:
Update README

gudnimg commented:
TODO
- [x] Add `--file <file>` argument to `update-po.py`
- [x] Add `--all` argument to `update-po.py`

gudnimg committed:
Add mutually exclusive arguments --file and --all

Only one of these arguments must be supplied, or argparse raises an error.

gudnimg committed:
Adjust module docstring

wavexx commented on lang/lang-extract.py:
```diff
@@ -261,13 +274,25 @@ def main():
 
+    for path in args.file:
+        if not Path(path).exists():
+            # assume its regex, search for files that match pattern
```
I assume you mean "glob" here?

wavexx commented on lang/lang-extract.py:
```diff
@@ -261,13 +274,25 @@ def main():
+            for file in sorted(PO_DIR.glob(path)):
+                FILE_LIST.append(file)
+        else:
+            FILE_LIST.append(Path(path))
```
Overall, since we now don't need the shell script, I would like to do this in ``update-pot.py`` from a style perspective.
No unix tool accepts/processes a glob itself normally, it's always done by calling program.

gudnimg commented:
* Rebased branch with respect to latest MK3 branch ✔️ 
* Dropped commit which deleted the .sh scripts ✔️

gudnimg committed:
Expand glob in update-pot on Windows only

gudnimg committed:
ubuntu: subprocess does not allow extended glob

It's my understanding that we cannot use extended globbing
on Linux because
of the way I am invoking the lang-extract
script.  the python script is not run through bash/shell.

We would need to perhaps use Shell = True in check_call() but
then the all the input arguments needs to be one string.

This commit was tested on Ubuntu 22.04.1 LTS

gudnimg commented:
@wavexx I've moved the glob extending to `update-pot.py`

Unfortunately we cannot use the built-in extended globbing on Linux because of the way I am invoking the `lang-extract.py` script. I tried to make it work. IMO it just makes the script more complicated. The script works as is at the moment. I tested on both Windows 10 and Ubuntu 22.04

If we really want to use extended globbing, I suggest we create a separate PR for it.

gudnimg commented:
Synced branch with latest MK3

TODO list:

- [x] Add a backup of `.po` files to be consistant with the shell script behavior.
- [x] Fix an issue where `update-po.py` doesn't add new messages from the `.pot` file. (Cannot reproduce on Windows 10 or Ubuntu 22)

For refence these are the new strings in the pot file which polib detect correctly:

```
#. MSG_TITLE_FILAMENT_EJECTED c=20
#: ../../Firmware/mmu2/errors_list.h:171 ../../Firmware/mmu2/errors_list.h:215
msgid "FILAMENT EJECTED"
msgstr ""

#. MSG_DESC_UNLOAD_MANUALLY c=20 r=8
#: ../../Firmware/mmu2/errors_list.h:258 ../../Firmware/mmu2/errors_list.h:310
msgid ""
"Filament detected unexpectedly. Ensure no filament is loaded. Check the "
"sensors and wiring."
msgstr ""

#. MSG_TITLE_LOAD_TO_EXTRUDER_FAILED c=20
#: ../../Firmware/mmu2/errors_list.h:139 ../../Firmware/mmu2/errors_list.h:181
msgid "LOAD TO EXTR. FAILED"
msgstr ""

#. MSG_DESC_LOAD_TO_EXTRUDER_FAILED c=20 r=8
#: ../../Firmware/mmu2/errors_list.h:227 ../../Firmware/mmu2/errors_list.h:277
msgid ""
"Loading to extruder failed. Inspect the filament tip shape. Refine the "
"sensor calibration, if needed."
msgstr ""

#. MSG_MATERIAL_CHANGES c=18
#: ../../Firmware/messages.cpp:103 ../../Firmware/ultralcd.cpp:1121
#: ../../Firmware/ultralcd.cpp:1195
msgid "Material changes"
msgstr ""

#. MSG_DESC_FILAMENT_EJECTED c=20 r=8
#: ../../Firmware/mmu2/errors_list.h:259 ../../Firmware/mmu2/errors_list.h:311
msgid "Remove the ejected filament from the front of the MMU unit."
msgstr ""

#. MSG_BTN_RESTART_MMU c=8
#: ../../Firmware/mmu2/errors_list.h:324 ../../Firmware/mmu2/errors_list.h:334
msgid "RstMMU"
msgstr ""

#. MSG_DESC_INSPECT_FINDA c=20 r=8
#: ../../Firmware/mmu2/errors_list.h:226 ../../Firmware/mmu2/errors_list.h:276
msgid ""
"Selector can't move due to FINDA detecting a filament. Make sure no filament"
" is in selector and FINDA works properly."
msgstr ""

#. MSG_DESC_FW_UPDATE_NEEDED c=20 r=9
#: ../../Firmware/mmu2/errors_list.h:262 ../../Firmware/mmu2/errors_list.h:308
msgid ""
"The MMU unit firmware version incompatible with the printer's FW. Update to "
"version 2.1.6."
msgstr ""
```

gudnimg committed:
Add a backup of po files

gudnimg committed:
Fix an issue with using --file argument

The following command did not work in the lang folder:
python .\update-po.py --file ./po/Firmware_cs.po

To resolve this we just extract the filename from
the path and use the constant PO_DIR to make
sure the path is always the same.

3d-gussner reviewed (approved):
Tested on Windows and Linux and works well.

Only minor difference between `.py` and `.sh` is the text warping, but nothing that causes issue in the firmware.

gudnimg committed:
Add executable permission for update-po.py and update-pot.py on Linux
